grpc_adapter.go

```diff
@@ -0,0 +1,668 @@
+		setHeaderCanonical(header, grpcHeaderAcceptCompression, acceptEncoding)
+	}
+
+	generateGetBody := func() {
+		request.Method = http.MethodPost
+		query := request.URL.Query()
+		request.URL.RawQuery = "" // clear query parameters
+
+		compressed := false
+		if contentEncoding := query.Get(connectUnaryCompressionQueryParameter); len(contentEncoding) > 0 {
+			compressed = contentEncoding != "identity"
+			delHeaderCanonical(header, connectUnaryHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+
+		msg := query.Get(connectUnaryMessageQueryParameter)
+		msgReader := queryValueReader(msg, query.Get(connectUnaryBase64QueryParameter) == "1")
+		request.Body = &bufferedEnvelopeReader{
+			ReadCloser: io.NopCloser(msgReader),
+			compressed: compressed,
+			config:     config,
+		}
+	}
+	generatePostBody := func() {
+		compressed := false
+		if contentEncoding := getHeaderCanonical(header, connectUnaryHeaderCompression); len(contentEncoding) > 0 {
+			compressed = contentEncoding != "identity"
+			delHeaderCanonical(header, connectUnaryHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+		request.Body = &bufferedEnvelopeReader{
+			ReadCloser: request.Body,
+			compressed: compressed,
+			config:     config,
+		}
+	}
+
+	// Handle GET request with query parameters
+	if request.Method == http.MethodGet {
+		generateGetBody()
+	} else {
+		generatePostBody()
+	}
+}
+
+func isProtocolHeader(header string) bool {
+	return header == connectUnaryHeaderCompression ||
+		header == connectUnaryHeaderAcceptCompression ||
+		header == connectStreamingHeaderCompression ||
+		header == connectStreamingHeaderAcceptCompression ||
+		header == connectHeaderTimeout ||
+		header == connectHeaderProtocolVersion ||
+		header == grpcHeaderCompression ||
+		header == grpcHeaderAcceptCompression ||
+		header == grpcHeaderTimeout ||
+		header == grpcHeaderStatus ||
+		header == grpcHeaderMessage ||
+		header == grpcHeaderDetails
+}
+
+type connectResponseWriter struct {
+	http.ResponseWriter
+	typ, enc string
+	config   *grpcAdapterConfig
+
+	body        bytes.Buffer // buffered body for unary payloads
+	header      http.Header  // buffered header for trailer capture
+	wroteHeader bool         // whether header has been written
+}
+
+func newConnectResponseWriter(responseWriter http.ResponseWriter, typ, enc string, config *grpcAdapterConfig) *connectResponseWriter {
+	return &connectResponseWriter{
+		ResponseWriter: responseWriter,
+		typ:            typ,
+		enc:            enc,
+		config:         config,
+	}
+}
+
+func (w *connectResponseWriter) Header() http.Header {
+	if w.header == nil {
+		w.header = make(http.Header)
+	}
+	return w.header
+}
+
+// isStreaming checks via the content type if the response is streaming.
+func (w *connectResponseWriter) isStreaming() bool {
+	return w.typ == connectStreamingContentTypeDefault
+}
```

Comment:
I think the implementation would be a little more straight-forward to split this into two response writers -- one for the unary protocol and one for the streaming protocol.